### PR TITLE
Translate SPIR-V OpUndef to LLVM "freeze poison"

### PR DESCRIPTION
**Pull request by jayfoad**
Translating OpUndef to poison causes CTS failures because poison is too undefined.

### CODE REVIEW

**Comment by jayfoad** [comment]
This fixes failures in dEQP-VK.spirv_assembly.instruction.graphics.opundef.matrix_* that were introduced by #2547.

**Comment by jayfoad** [comment]
This is more complicated than I realised. `freeze` is an instruction so it has to appear in a function body, but `OpUndef` is more like a constant so it can appear anywhere. I can think of a couple of options:
1. For each _use_ of `OpUndef`, if it is in a function, insert a `freeze poison` at the start of that function.
2. Go back to translating `OpUndef` to LLVM `undef` (witha FIXME comment).

Maybe we need a better understanding of exactly why `poison` was causing CTS failures.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 915cdbe1b53253ddd32b2d7bcc118415a939e016
<details><summary>CTS tests (Failed: 0/211353)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 35147/70456 (49.9%)</li>
<li>Failed: 0/70456 (0.0%)</li>
<li>Not Supported: 35309/70456 (50.1%)</li><li>Warnings: 0/70456 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39026/70456 (55.4%)</li>
<li>Failed: 0/70456 (0.0%)</li>
<li>Not Supported: 31430/70456 (44.6%)</li><li>Warnings: 0/70456 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39058/70441 (55.4%)</li>
<li>Failed: 0/70441 (0.0%)</li>
<li>Not Supported: 31383/70441 (44.6%)</li><li>Warnings: 0/70441 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by jayfoad** [comment]
> Maybe we need a better understanding of exactly why `poison` was causing CTS failures.

The tests check that OpUndef * 0.0 results in either 0.0 or NaN. In LLVM IR that is true for undef, but poison * 0.0 is poison so it fails.

**Comment by dstutt** [comment]
So what does it emit as a final result?

**Comment by jayfoad** [comment]
> This fixes failures in dEQP-VK.spirv_assembly.instruction.graphics.opundef.matrix_* that were introduced by #2547.

The updated patch is now just a partial revert of #2547, plus a TODO comment.

**Comment by amdvlk-admin** [comment]
### Test summary for commit f8b6be2836d52762e767283b6b28112fc47e64c5
<details><summary>CTS tests (Failed: 0/211353)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 35147/70456 (49.9%)</li>
<li>Failed: 0/70456 (0.0%)</li>
<li>Not Supported: 35309/70456 (50.1%)</li><li>Warnings: 0/70456 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39026/70456 (55.4%)</li>
<li>Failed: 0/70456 (0.0%)</li>
<li>Not Supported: 31430/70456 (44.6%)</li><li>Warnings: 0/70456 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39058/70441 (55.4%)</li>
<li>Failed: 0/70441 (0.0%)</li>
<li>Not Supported: 31383/70441 (44.6%)</li><li>Warnings: 0/70441 (0.0%)</li>
</ul></details>


</ul></details>

**Review by dstutt — APPROVED** [review]
This LGTM to fix the CTS failures.

You might want to wait for reviews from Rex or Ruiling though?

**Review by amdrexu — APPROVED** [review]

**Pull request merged**
